crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2720,39 +2680,80 @@ impl World {
         let last_change_tick = self.last_change_tick();
         let change_tick = self.change_tick();
 
-        let component_id = self.components.get_valid_resource_id(TypeId::of::<R>())?;
-        let (ptr, mut ticks, mut caller) = self
-            .storages
-            .resources
-            .get_mut(component_id)
-            .and_then(ResourceData::remove)?;
-        // Read the value onto the stack to avoid potential mut aliasing.
-        // SAFETY: `ptr` was obtained from the TypeId of `R`.
-        let mut value = unsafe { ptr.read::<R>() };
+        let component_id = self.components.valid_resource_id::<R>()?;
+        let entity = *self.resource_entities.get(component_id)?;
+        let mut entity_mut = self.get_entity_mut(entity).ok()?;
+
+        let mut ticks = entity_mut.get_change_ticks::<R>()?;
+        let mut changed_by = entity_mut.get_changed_by::<R>()?;
+        let mut value = entity_mut.take::<R>()?;
```

Comment:
try_resource_scope already had a nasty performance cost for what should be a simple access. Introducing archetype + table moves, component hook execution, etc into the equation makes this considerably less defensible.

resource_scope was already a bit of a UX nightmare (see the many nasted scopes in some of the bevy internals). It was "barely good enough", and with this implementation I consider it "not good enough" for the current use cases. We should discuss what the future of this type of access looks like, because removing every piece of data we want to access is an increasingly absurd approach.